core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -144,8 +164,24 @@ public long acquireExecutionMemory(long required, MemoryConsumer consumer) {
       // spilling, avoid to have too many spilled files.
       if (got < required) {
         // Call spill() on other consumers to release memory
+        // Sort the consumers according their memory usage. So we avoid spilling the same consumer
+        // which is just spilled in last few times and re-spilling on it will produce many small
+        // spill files.
+        List<MemoryConsumer> sortedList = new ArrayList<>(consumers.size());
         for (MemoryConsumer c: consumers) {
           if (c != consumer && c.getUsed() > 0 && c.getMode() == mode) {
+            sortedList.add(c);
+          }
+        }
+        Collections.sort(sortedList, new ConsumerComparator());
+        for (int listIndex = 0; listIndex < sortedList.size(); listIndex++) {
+          MemoryConsumer c = sortedList.get(listIndex);
+          // Try to only spill on the consumer which has the required size of memory.
+          // As the consumers are sorted in descending order, if the next consumer doesn't have
+          // the required memory, then we need to spill the current consumer at least.
+          boolean doSpill = (listIndex + 1) == sortedList.size() ||
+            sortedList.get(listIndex + 1).getUsed() < (required - got);
+          if (doSpill) {
```

Comment:
I like the fact that this implementation does not need to incur the cost of remove in a TreeMap.
Unfortunately, I dont think it is sufficient : the impl assumes that spill() will actually always give you back getUsed - from the rest of the code in the method, this does not look like a valid assumption to make.

This can resulting in spilling a large number of smaller blocks, and potentially itself.

For example: required = 500MB, consumers = 1.5GB 1GB 500MB 2MB 1MB ..
If spilling 500MB resulted in (say) releasing 490MB, we might end up spilling a large number of blocks and also (potentially) end up spilling itself - also can end up returning less than requested while enough memory does exist to satisfy the request.